tracing/ops_tracing/_buffer.py

```diff
@@ -0,0 +1,291 @@
+                exc = e
+                continue
+        else:
+            assert exc  # noqa: S101  # we'd have returned otherwise
+            raise exc
+
+    return wrapper
+
+
+class Buffer:
+    """Buffer for tracing data.
+
+    Access to the buffer attributes is effectively protected by an sqlite transaction.
+    """
+
+    ids: set[int]
+    """Tracing data ids buffered during this dispatch invocation."""
+    observed = False
+    """Marks that data from this dispatch invocation has been marked observed."""
+    stored: int | None = None
+    """Estimate of stored data size in bytes if known, or None if not yet known."""
+
+    def __init__(self, path: pathlib.Path | str):
+        self.path = str(path)
+        self.ids = set()
+        self._set_db_schema()
+
+    @retry
+    def _set_db_schema(self):
+        # TODO: measure the cost of this vs two-level approach:
+        # - check table and index in read-only mode
+        # - if needed, update the DSL
+        # Note that ops storage sets u+rw, go-rw permissions
+        # should we follow suit?
+        with self.tx(timeout=LONG_DB_TIMEOUT) as conn:
+            conn.execute(
+                """
+                CREATE TABLE IF NOT EXISTS tracing (
+                    -- effectively auto-incremented
+                    id INTEGER PRIMARY KEY,
+                    -- observed events are more important
+                    priority INTEGER NOT NULL,
+                    -- a chunk of tracing data
+                    data BLOB NOT NULL,
+                    -- MIME type for these data
+                    mime TEXT NOT NULL
+                )
+                """
+            )
+            conn.execute(
+                """
+                CREATE INDEX IF NOT EXISTS tracing_priority_id
+                ON tracing
+                (priority, id)
+                """
+            )
+            conn.execute(
+                """
+                CREATE TABLE IF NOT EXISTS settings (
+                    -- setting key, like "url" and "ca"
+                    key TEXT PRIMARY KEY,
+                    -- the value, with None coerced to empty string
+                    value TEXT NOT NULL
+                )
+                """
+            )
+
+    @contextlib.contextmanager
+    def tx(self, *, timeout: float = DB_TIMEOUT, readonly: bool = False):
+        """Thread-safe transaction context manager."""
+        with sqlite3.connect(self.path, isolation_level=None, timeout=timeout) as conn:
+            mode = 'DEFERRED' if readonly else 'IMMEDIATE'
+            conn.execute(f'BEGIN {mode}')
+            try:
+                yield conn
+            except:
+                conn.execute('ROLLBACK')
+                raise
+            else:
+                conn.execute('COMMIT')
+
+    @retry
+    def get_destination(self) -> Config:
+        """Get the tracing destination from the database."""
+        with self.tx(readonly=True) as conn:
```

Comment:
Why is a transaction necessary for a read?

Reply:
True, I should clean up all the single-statement functions.

Reply:
At the same time, I wanted to isolate `with sqlite3.connect(<fancy args>)` into a single function, to make sure that the arguments are consistent.